MAGE Dark Mode 1.0

ConConner opened the pull request:
## Adds full theming capabilities to MAGE.  
Due to the nature of WinForms this required the use of some custom controls and weird workarounds.
* Forms have special code that initializes the theme
* Minor adjustments on some forms
* Replacement of stock controls on each form with custom controls

Custom controls include:
* Flat TextBox
* Flat ComboBox
* Flat TabControl
* Flat NumericUpDown

A new Form exists, the ThemeEditor, which allows for the creation of custom themes.  
There are two standard themes: "Mage Old" and "Mage Dark", which cannot be deleted.  
Themes can be exported and imported as `MAGE Theme Files (.mtf)`.  

The custom code that is not integrated into Forms directly is relatively encapsulated in a new namespace called `Theming`

## Known Bugs
- List View coloumn headers are not themed
- Tree View structure is not themed
- Text Boxes have very intense flickering when hovering over them
  - This is a very tough one to fix
- Scroll bars are not themed
  - pretty much impossible to fix. Would require a custom scroll bar control
  - said control would need to be injected in every control that can display them, somehow

ConConner committed:
Basics of dark mode

Probably more complicated than expected. Especially scroll bars and other Win Api wrapper controls need to be changed or remade

ConConner committed:
Custom flat combobox

Comboboxes are tricky, they will require a custom control. Before replacing them I want to back up though

ConConner committed:
Panic push

ConConner committed:
First Milestone

Menu is fully customizable now.
Next up
- Check boxes
- Text boxes
- Scroll bars

Scroll bars are gonna be really bad

ConConner committed:
Check boxes

Check boxes work.
Scroll bars are still left
other forms need support

ConConner committed:
Theme Switching

Themes can be switched in the Theme Editor. Edited colors cannot be saved though

ConConner committed:
Color theme changing

Color themes can be changed now. New ones cannot be added and current edits cannot be saved

ConConner committed:
FlatNumUpDown

Added a flat NumericUpDown
changed some more editors.
ListView will require some heavy reword

ConConner committed:
Injected Theming into all Forms

Theming is now on all Forms.
Some controls still need a rework.

ConConner committed:
Finishing Touches

Finishing Touches for a test.
There are still some very nasty controls
and bugs left but it is pretty usable so far.

biosp4rk commented on mage/FormShortcuts.cs:
```diff
@@ -14,6 +15,12 @@ public FormShortcuts(FormMain main)
+            ThemeSwitcher.ChangeTheme(Controls, this);
+            ThemeSwitcher.InjectPaintOverrides(Controls);
+
+            ThemeSwitcher.ChangeTheme(this.Controls, this);
+            ThemeSwitcher.InjectPaintOverrides(this.Controls);
+
```
Looks like this is unnecessary

ConConner replied:
No this is necessary in every form to do the actual theme changing

biosp4rk replied:
Look at the full context, it's a duplicate

ConConner replied:
Oh you are right

biosp4rk commented on mage/Theming/ThemeSwitcher.cs:
```diff
@@ -0,0 +1,327 @@
+            {
+                if (!Themes.ContainsKey(value))
+                {
+                    if (value == "Mage Old") Themes.Add("Mage Old", StandardTheme);
```
You should add constants for theme names like "Mage Old" and "Mage Dark"

biosp4rk commented on mage/Editors/FormPalette.cs:
```diff
@@ -294,21 +301,43 @@ private unsafe void DrawColorBar()
         private unsafe void DrawLine(BitmapData imgData, int color)
```
Is it possible to call this function if the palette editor is open when the theme changes?

ConConner committed:
Small fixes

Changes to address biosparks comments

ConConner committed:
csproj revert

reverted csproj file to remove the weird vs changes

ConConner committed:
Merge pull request #1 from ConConner/dev

csproj revert

ConConner committed:
.NET 6 switch

ConConner committed:
Fixed Minimap editor

Minimap Editor is fixed.
Also changed the link color in the about page

ConConner committed:
Merge pull request #2 from ConConner/dev-framework-switch

Dev framework switch

ConConner committed:
New TextBox

Changed TextBox again, finally no flicker and everything seems to work properly.
Added code to give a text color back with the highest contrast to the accent color

ConConner committed:
contrast text in menu bar

added the contrasted text in the menu strip

ConConner committed:
Text Box fix

Text box text evernt works again

ConConner committed:
Theme editor working again

Theme editor works

ConConner committed:
Form size fix

Forms are no longer inflated

ConConner committed:
Minimap Fix No.100

Maybe this one will be the last?

ConConner committed:
Removed Newtonsoft.JSON

Switched to the built in JSON library.
Added custom converter to allow Color serialization and keep compatibility with the old themes.

ConConner committed:
Bulk import

Themes can be imported from a zip file

ConConner committed:
Added Bulk export

Added bulk export option

ConConner committed:
Fixed weird link bug

This probably happened due to the Framework switch but it is now required to specifiy UseShellExecute as true

ConConner committed:
Sprite Editor fix

Sprite Editor would not display Total drops correctly
